### PR TITLE
Creating a PySpark DataFrame: A Beginner's Guide

### LINKED ISSUE #5003: [Languages]Creating a PySpark DataFrame: A Beginner's Guide
### Topic Suggestion
Creating a PySpark DataFrame: A Beginner's Guide
#### Proposed article introduction
We can distribute data and conduct calculations on several nodes of a cluster using Spark, a cluster computing platform. It is easier to process large datasets when the data is distributed.
In this example, each node is described as a discrete computer that is focused on a certain subset of data. This node would also be responsible for some of the calculations that take place during dataset operations. In addition to Scala, Spark also supports Java, Python, R and SQL programming languages.

#### Key takeaways 
- PySpark DataFrame From an Existing RDD
- PySpark DataFrame From an External File
- Additional Useful Method

#### Article quality
In this lesson, we will learn about creating PySpark DataFrames. PySpark DataFrames and two techniques for creating them will be covered in this article. This article differs from others in that I'll generate a PySpark DataFrame from an existing RDD, generate PySpark DataFrame from List and from external file sources using Google Colaboratory for practice. As a bonus, we'll cover certain approaches that aren't already covered in the published papers on PySpark. The article will be comprehensive, and the code provided will be simple to grasp.
### Reference
N/A
### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by FranciscaNg**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #5003 

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by FranciscaNg** [comment]
@Bayler Please look at this PR whenever you are free. Thanks

**Comment by mohamedgh16** [comment]
@hectorkambow I would like to review this article.

**Comment by mohamedgh16** [comment]
Great article @FranciscaNg.
Follow my suggestions and implement the changes so we can move on to the tech review.

**Review by ghost — COMMENTED** [review]
Great article @FranciscaNg.
Follow my suggestions and implement the changes so we can move on to the tech review.

**Comment by FranciscaNg** [comment]
@mohamedgh16 
I will. Thanks

File: content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+The `resilientDistributedDataset` data type will determine RDD creation success or failure. We should get `pyspark.resilientDistributedDataset.RDD` as a result. This is seen in the code below:
+```Python
+print(type(resilientDistributedDataset))
+```
+Output:
+```bash
+<class 'pyspark.rdd.RDD'>
+```
+#### 5. Convert the Resilient Distributed Dataset into PySpark DataFrame
```
Comment (mohamedgh16) [inline content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md:51]: Use the term RDD.

File: content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+```python
+subjects = ['Division','Kiswahili','Biology','Agriculture','Mathematics']
+ourMarks_df = sparkSession.createDataFrame(resilientDistributedDataset, schema=subjects)
+```
+In this case, the RDD is sent in the form of data. The schema parameter of the `createDataFrame()` method will be filled with previously produced string topics.
```
Comment (mohamedgh16) [inline content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md:56]: In this case, the RDD is sent as data. The schema parameter of the `createDataFrame()` method will be filled with previously produced string topics.

File: content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+We'll look at the data type of `ourMarks_df` below to see if everything went well.
```
Comment (mohamedgh16) [inline content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md:58]: We'll look at the data type of `ourMarks_df` below to see if everything working correctly.

File: content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+We'll look at the data type of `ourMarks_df` below to see if everything went well. 
+```Python
+print(type(ourMarks_df))
+```
+Output:
+```bash
+<class 'pyspark.sql.dataframe.DataFrame'>
+```
+Using the `printSchema()` method, we can analyze our schema, which is useful when we have many columns.
```
Comment (mohamedgh16) [inline content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md:66]: Using the `printSchema()` method, we can analyze the schema, which is useful when we have many columns.

File: content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
++--------+---------+-------+-----------+-----------+
+|       C|       80|     70|         85|         91|
+|       B|       80|     70|         85|         91|
+|       A|       80|     78|         96|         92|
+|       A|       92|     70|         89|         96|
++--------+---------+-------+-----------+-----------+
+```
+> To see the whole code for this [click here](https://colab.research.google.com/drive/1Ykwifphdj2mLQ9rhKuX309aXq-tGurTD?usp=sharing)
+### Creating Python data frame from an external file using PySpark
+Our external Files will be imported using the spark session's `read()` function. Data frame object will be created as the result of this action. The techniques to import these file formats are practically comparable, and one may import them with minimal effort. In contrast to the previous method, this one is simple and requires a spark session to produce a PySpark data frame from a Resilient Distributed Dataset.
```
Comment (mohamedgh16) [inline content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md:97]: The external files will be imported using the spark session's `read()` function. A data frame object will be created as a result of this action. The techniques to import these file formats are practically comparable, and one may import them with minimal effort.

 In contrast to the previous method, this one is simple and requires a spark session to produce a PySpark data frame from a Resilient Distributed Dataset.

File: content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+#### 1. Importing the libraries:
+```Python
+from pyspark.sql import SparkSession
+```
+#### 2. Creating the spark session
+```python
+sparkSession = SparkSession.builder.appName('Creating PySpark dataframe from an external file').getOrCreate()
+```
+Using `appName(),` we can retrieve the application's name by passing in a string. It is possible to leverage an existing spark session or build a new one using the `getOrCreate()` method.
```
Comment (mohamedgh16) [inline content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md:107]: Using `appName(),` we can retrieve the application's name by passing a string.

File: content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+#### 3. Read the external files into the PySpark data frame
+The external file format that we will use and import comprises a CSV file.
+
+```python
+ourCSV_file = sparkSession.read.csv('IncomeAndGender.csv', sep = ',', inferSchema = True, header = True)
+```
+Delimiters for the CSV file were given (which is a comma). Afterward, the inferSchema property was set to True. CSV files will be automatically transformed into PySpark data frames with this command.
```
Comment (mohamedgh16) [inline content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md:114]: Delimiters for the CSV file are given (as a comma). Afterward, the inferSchema property was set to True. CSV files will be automatically transformed into PySpark data frames with this command.

File: content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+Using `appName(),` we can retrieve the application's name by passing in a string. It is possible to leverage an existing spark session or build a new one using the `getOrCreate()` method.
+#### 3. Read the external files into the PySpark data frame
+The external file format that we will use and import comprises a CSV file.
+
+```python
+ourCSV_file = sparkSession.read.csv('IncomeAndGender.csv', sep = ',', inferSchema = True, header = True)
+```
+Delimiters for the CSV file were given (which is a comma). Afterward, the inferSchema property was set to True. CSV files will be automatically transformed into PySpark data frames with this command. 
+
+Let's have a look at our PySpark data frame's data kinds now:
```
Comment (mohamedgh16) [inline content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md:116]: Let's have a look at the PySpark data frame's data kinds:

File: content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+```
+> To see the complete code for this [click here](https://colab.research.google.com/drive/1u04j9muZ1FdDvr2oxeuSGST5Lva9es9j?usp=sharing). In addition, the CSV file that we used here was initially created in our notebook.
+
+### Other methods 
+#### 1. Converting PySpark data frame to a Pandas data frame
+The PySpark data frame can be converted to a Pandas data frame, and Pandas techniques can be used on our data frame, which is incredibly convenient.
+```python
+csvToPandas = ourCSV_file.toPandas()
+```
+To ensure that everything we're doing is correct, let's look at the data type of our new data frame.
```
Comment (mohamedgh16) [inline content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md:144]: To ensure that everything works correctly, let's look at the data type of our new data frame.

File: content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+pandas.core.frame.DataFrame
+```
+#### 2. Reading multiple files
+In the `read()` method, we can pass a list of file paths as a string type, creating and allocating a PySpark data frame in the variable `csvToPandas.` A shared data frame should not be deemed significant while reading several files at the same time.
+### Reference
+1. See full code for [creating PySpark data frame from an external file](https://colab.research.google.com/drive/1u04j9muZ1FdDvr2oxeuSGST5Lva9es9j?usp=sharing).
+2. See whole code for [creating a PySpark data frame using an existing Resilient Distributed Dataset](https://colab.research.google.com/drive/1Ykwifphdj2mLQ9rhKuX309aXq-tGurTD?usp=sharing)
+3. [pyspark.sql.DataFrame](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.DataFrame.html)
+### Conclusion
+Compared to `read()` methods, `parallelize()` is better, but it necessitates more work. When we need to fast scan a CSV file, the `read()` methods are helpful. Many options are available when making changes to a PySpark data frame.
```
Comment (mohamedgh16) [inline content/articles/creating-a-pyspark-data-frame-a-beginners-guide/index.md:159]: Use the conclusion to summarize the main ideas that you explained through this article, add more value to it.

**Review by ghost — COMMENTED** [review]
@FranciscaNg let me know when you're done.

**Comment by FranciscaNg** [comment]
In the next few minutes I'll be done. Thanks

**Comment by FranciscaNg** [comment]
@mohamedgh16 Done

**Comment by mohamedgh16** [comment]
@ahmadmardeni1 this article is ready for final review.

**Comment by mohamedgh16** [comment]
Front matter
---
layout: engineering-education
status: publish
published: true
url: /creating-a-pyspark-dataframe-a-beginners-guide/
title: Creating a PySpark DataFrame: A Beginner's Guide
description: This article will guide you to create PySpark data frames .
author: francisca-ngodu
date: 2021-12-26
topics: [Languages]
excerpt_separator: 
images:

  - url: /engineering-education/creating-a-pyspark-dataframe-a-beginners-guide/hero.jpg
    alt: Creating a PySpark DataFrame: A Beginner's Guide Cover image
---

Peer Review Contributions by: [Mohamed Alghadban](/engineering-education/authors/mohamed-alghadban/)

**Comment by ahmadmardeni1** [comment]
It didn't pass the plagiarism check yet. @mohamedgh16 
Cc @LinusMuema

**Comment by WanjaMIKE** [comment]
I found this article to be a paraphrase of this one https://www.analyticsvidhya.com/blog/2021/09/beginners-guide-to-create-pyspark-dataframe/ Only a few things have been changed. We usually encourage contributors to submit unique topics and PRs to provide more value to the community. I think this PR cannot be fixed and thus subject to closure..

**Comment by ahmadmardeni1** [comment]
Please choose another PR, and @ me. @mohamedgh16

**Pull request closed without merging**